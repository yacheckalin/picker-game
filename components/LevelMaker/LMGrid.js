import React, { useState, useEffect, useRef } from "react";

import PropTypes from "prop-types";
import styled from "styled-components";

import {
  DEFAULT_CELL_SIZE,
  MARK_SELECTED,
  MARK_CLEAR,
  KEYS,
  DOORS,
  WALL,
  WALL_D,
  EMPTY,
} from "../../constants";

import LMGridTools from "./LMGridTools";

import shortid from "shortid";
//TODO: change the colors here
const StyledColumn = styled.div`
  margin: 0px;
  padding: 0px;
  width: ${(prop) => prop.width}px;
  height: ${(prop) => prop.height}px;
  border: 1px solid #bdbdbd;
  display: inline-flexbox;

  label:hover {
    background-color: #ffccbc;
  }
  input {
    display: none;
  }
  input:checked + label {
    background-color: #ff8a65;
  }
  label {
    display: block;
    width: 100%;
    height: 100%;
    background-color: #c5e1a5;
  }
  label > i {
    display: block;
    width: 100%;
    height: 100%;
    font-size: 2em;
    color: red;
  }
`;

const StyledRow = styled.div`
  margin: 0px;
  padding: 0px;
`;

const StyledContainer = styled.div`
  position: relative;
`;

const LMGrid = ({ data, size }) => {
  const containerRef = useRef();

  const [cellWidth, setCellWidth] = useState(DEFAULT_CELL_SIZE);
  const [cellHeight, setCellHeight] = useState(DEFAULT_CELL_SIZE);
  const [cellTop, setCellTop] = useState(0);
  const [cellLeft, setCellLeft] = useState(0);
  const [openTool, setOpenTool] = useState(false);
  const [gridWidth, setGridWidth] = useState(0);
  const [isMultipleSelect, setIsMultipleSelect] = useState(false);
  const [gridData, setGridData] = useState(data);

  useEffect(() => {
    const { offsetWidth } = containerRef.current;
    const cellSize = Math.floor(parseInt(offsetWidth) / size) - 1;

    setGridWidth(offsetWidth);
    setCellHeight(cellSize);
    setCellWidth(cellSize);
  }, [cellWidth, cellHeight, gridWidth]);

  const handleGridMarking = (e) => {
    const { offsetTop, offsetLeft } = e.target;

    setCellTop(offsetTop);
    setCellLeft(offsetLeft);
    setOpenTool(false);

    const tmp = e.target.id.split("-");
    if (tmp) {
      const [y, x] = tmp.splice(-2, 2);

      if (x >= 0 && y >= 0) {
        gridData[y][x] = e.target.checked ? MARK_SELECTED : MARK_CLEAR;
        setGridData(gridData);
      }
    }
  };

  const mapObjectTagWithKey = (tag) => {
    if (KEYS[tag]) return KEYS[tag];
    if (DOORS[tag]) return DOORS[tag];
    if (tag === "WALL") return WALL;
    if (tag === "WALL_D") return WALL_D;
  };

  const handleInsert = ({ key }) => {
    for (let i = 0; i < gridData.length; i++) {
      for (let j = 0; j < gridData[i].length; j++) {
        if (gridData[i][j] === MARK_SELECTED) {
          gridData[i][j] = mapObjectTagWithKey(key);
        }
      }
    }

    setGridData(gridData);
    setOpenTool(false);
  };

  const keypressHandler = (e) => {
    if (e.key == "Enter") {
      // check if there more then one cell selected

      let selectedCells = [];
      for (const row of data) {
        for (const col of row) {
          if (col === MARK_SELECTED) {
            selectedCells.push(col);
          }
        }
      }
      setIsMultipleSelect(selectedCells.length > 1 ? true : false);
      setOpenTool(true);
    }
  };
  useEffect(() => {
    window.addEventListener("keypress", keypressHandler);
    return () => {
      window.removeEventListener("keypress", keypressHandler);
    };
  });

  const renderIcon = (tag) => {
    switch (tag) {
      case WALL:
        return `indigo`;
      case WALL_D:
        return "indigo lighten-3";
      case DOORS.RED_DOOR:
        return "red darken-3";
      case DOORS.BLUE_DOOR:
        return "blue darken-3";
      case DOORS.GREEN_DOOR:
        return "green darken-3";
      case KEYS.RED_KEY:
        return "red";
      case KEYS.BLUE_KEY:
        return "blue";
      case KEYS.GREEN_KEY:
        return "green";
    }
  };

  return (
    <StyledContainer className="row" ref={containerRef}>
      <div className="col s12 ">
        {gridData.map((row, y) => (
          <StyledRow key={y} className="row">
            {row.map((col, x) => (
              <StyledColumn
                width={cellWidth}
                height={cellHeight}
                key={x}
                onClick={handleGridMarking}
                id={`lm-grid-cell-${y}-${x}`}
              >
                <input type="checkbox" id={`lm-grid-cell-input-${y}-${x}`} />
                <label
                  htmlFor={`lm-grid-cell-input-${y}-${x}`}
                  className={renderIcon(col)}
                ></label>
              </StyledColumn>
            ))}
          </StyledRow>
        ))}

        <LMGridTools
          open={openTool}
          cellWidth={cellWidth}
          cellHeight={cellHeight}
          cellSize={size}
          parentTop={cellTop}
          parentLeft={cellLeft}
          gridWidth={gridWidth}
          key={shortid.generate()}
          multipleMode={isMultipleSelect}
          handleInsert={handleInsert}
        />
      </div>
    </StyledContainer>
  );
};

LMGrid.propTypes = {
  data: PropTypes.array.isRequired,
  size: PropTypes.number.isRequired,
};

export default LMGrid;
