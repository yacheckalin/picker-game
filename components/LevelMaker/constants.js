export const MIN_MAP_SIZE = 3;
export const MAX_MAP_SIZE = 50;
export const DEFAULT_GRID_MAP = [
  [0, 0, 0],
  [0, 0, 0],
  [0, 0, 0],
];

export const DEFAULT_CELL_SIZE = 40;
export const MARK_SELECTED = 999;
