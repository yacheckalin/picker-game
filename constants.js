export const KEYS = {
  BLUE_KEY: [20, "BLUE"],
  GREEN_KEY: [21, "GREEN"],
  RED_KEY: [22, "RED"],
};
export const DOORS = {
  BLUE_DOOR: [90, "BLUE"],
  GREEN_DOOR: [91, "GREEN"],
  RED_DOOR: [92, "RED"],
};
export const WALL = 1;
export const WALL_D = 2; // walls behind the door
export const VISITED = 8;
export const EMPTY = 0;
export const EMPTY_BACKPACK_CELL = 0;
export const DEFAULT_FIELD_SIZE = 4;
export const DEFAUTL_FIELD_HASH = "level_default";

export const MIN_MAP_SIZE = 3;
export const MAX_MAP_SIZE = 50;
export const DEFAULT_GRID_MAP = [
  [0, 0, 0],
  [0, 0, 0],
  [0, 0, 0],
];

export const DEFAULT_CELL_SIZE = 40;
export const MARK_SELECTED = 999;
export const MARK_CLEAR = 0;
