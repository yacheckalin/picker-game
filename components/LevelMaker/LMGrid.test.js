import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import LMGrid from "./LMGrid";

import { DEFAULT_GRID_MAP, DEFAULT_GRID_SIZE } from "../../constants";

jest.mock("shortid", () => {
  return {
    generate: jest.fn(() => 1),
  };
});

describe("<LMGrid />", () => {
  const props = {
    data: DEFAULT_GRID_MAP,
    size: DEFAULT_GRID_SIZE,
  };

  test("displays and shows properly", () => {
    const wrapper = shallow(<LMGrid {...props} />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
