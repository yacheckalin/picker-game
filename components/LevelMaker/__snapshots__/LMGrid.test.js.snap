// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LMGrid /> displays and shows properly 1`] = `
<styled.div
  className="row"
>
  <div
    className="col s12 "
  >
    <styled.div
      className="row"
      key="0"
    >
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-0-0"
        key="0"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-0-0"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-0-0"
        />
      </styled.div>
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-0-1"
        key="1"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-0-1"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-0-1"
        />
      </styled.div>
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-0-2"
        key="2"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-0-2"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-0-2"
        />
      </styled.div>
    </styled.div>
    <styled.div
      className="row"
      key="1"
    >
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-1-0"
        key="0"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-1-0"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-1-0"
        />
      </styled.div>
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-1-1"
        key="1"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-1-1"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-1-1"
        />
      </styled.div>
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-1-2"
        key="2"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-1-2"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-1-2"
        />
      </styled.div>
    </styled.div>
    <styled.div
      className="row"
      key="2"
    >
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-2-0"
        key="0"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-2-0"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-2-0"
        />
      </styled.div>
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-2-1"
        key="1"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-2-1"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-2-1"
        />
      </styled.div>
      <styled.div
        className="generated-grid-cell"
        height={40}
        id="lm-grid-cell-2-2"
        key="2"
        onClick={[Function]}
        width={40}
      >
        <input
          id="lm-grid-cell-input-2-2"
          type="checkbox"
        />
        <label
          htmlFor="lm-grid-cell-input-2-2"
        />
      </styled.div>
    </styled.div>
    <LMGridTools
      cellHeight={40}
      cellSize={3}
      cellWidth={40}
      gridWidth={0}
      handleInsert={[Function]}
      key="1"
      multipleMode={false}
      open={false}
      parentLeft={0}
      parentTop={0}
    />
  </div>
</styled.div>
`;
